src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -924,13 +877,18 @@ MultiTaskRouterSelectQuerySupported(Query *query)
 								 NULL, NULL);
 		}
 
-		/* see comment on AddUninstantiatedPartitionRestriction() */
 		if (subquery->setOperations != NULL)
 		{
-			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-								 "set operations are not allowed in INSERT ... SELECT "
-								 "queries",
-								 NULL, NULL);
+			SetOperationStmt *setOperationStatement =
+				(SetOperationStmt *) subquery->setOperations;
+
+			if (setOperationStatement->op != SETOP_UNION)
```

Comment:
You should revisit this when enabling UNIONS as there could me more than one set operation such ash  a U (b - c).